tests/utils.py

```diff
@@ -104,12 +104,31 @@ async def assert_revert_entry_point(fun, invalid_selector):
     await assert_revert(fun, entry_point_msg)
 
 
-def assert_event_emitted(tx_exec_info, from_address, name, data):
-    assert Event(
-        from_address=from_address,
-        keys=[get_selector_from_name(name)],
-        data=data,
-    ) in tx_exec_info.raw_events
+def assert_event_emitted(tx_exec_info, name, data, order=0):
+    """Assert one single event is fired with correct data."""
+    assert_events_emitted(tx_exec_info, [(order, name, data)])
+
+
+def assert_events_emitted(tx_exec_info, events):
+    """Assert events are fired with correct data."""
+    for event in events:
+        order, name, data = event
+        event_obj = OrderedEvent(
+            order=order,
+            keys=[get_selector_from_name(name)],
+            data=data,
+        )
+
+        if event_obj in tx_exec_info.call_info.internal_calls[0].events:
+            return
+
+        try:
+            if event_obj in tx_exec_info.call_info.internal_calls[0].internal_calls[0].events:
+                return
+        except IndexError:
```

Comment:
I believe it can nest even deeper depending on the amount of contract calls in the tx (I have to research further to see how far it could go). For our tests, the longer dot notation is for the proxy tests and the shorter is for the rest (I need to double check to confirm that's always the case). As mentioned in another comment, I'm assuming it will be easier to fetch events as before once the testing framework is updated. Otherwise, we can certainly consider how deeply we should look and adjust.

Reply:
i wonder why are you so sure it's temporal but if that's the case then it's ok!